upload-api/stores/blobs.js

```diff
@@ -0,0 +1,134 @@
+import { base64pad } from 'multiformats/bases/base64'
+import { decode as digestDecode } from 'multiformats/hashes/digest'
+import { base58btc } from 'multiformats/bases/base58'
+
+import {
+  S3Client,
+  HeadObjectCommand,
+  PutObjectCommand,
+} from '@aws-sdk/client-s3'
+import { getSignedUrl } from '@aws-sdk/s3-request-presigner'
+
+/**
+ * @typedef {import('@web3-storage/upload-api/types').BlobsStorage} BlobsStorage
+ * @typedef {import('@ucanto/interface').Failure} Failure
+ * @typedef {import('@ucanto/interface').Result<boolean, Failure>} HasResult
+ */
+
+/**
+ * Abstraction layer with Factory to perform operations on bucket storing Blobs.
+ *
+ * @param {string} region
+ * @param {string} bucketName
+ * @param {import('@aws-sdk/client-s3').ServiceInputTypes} [options]
+ */
+export function createBlobsStorage(region, bucketName, options) {
+  const s3 = new S3Client({
+    region,
+    ...options,
+  })
+  return useBlobsStorage(s3, bucketName)
+}
+
+/**
+ *
+ * @param {S3Client} s3
+ * @param {string} bucketName
+ * @returns {BlobsStorage}
+ */
+export function useBlobsStorage(s3, bucketName) {
+  return {
+    /**
+     * @param {Uint8Array} multihash
+     */
+    has: async (multihash) => {
+      const encodedMultihash = base58btc.encode(multihash)
+      const cmd = new HeadObjectCommand({
+        Key: `${encodedMultihash}/${encodedMultihash}.blob`,
+        Bucket: bucketName,
+      })
+      try {
+        await s3.send(cmd)
+      } catch (cause) {
+        // @ts-expect-error
+        if (cause?.$metadata?.httpStatusCode === 404) {
+          return { ok: false }
+        }
+        throw new Error('Failed to check if car-store', { cause })
+      }
+      return { ok: true }
+    },
+    /**
+     * Create a presigned s3 url allowing the recipient to upload
+     * only the CAR that matches the provided Link
+     *
+     * @param {Uint8Array} multihash
+     * @param {number} size
+     */
+    createUploadUrl: async (multihash, size) => {
+      const encodedMultihash = base58btc.encode(multihash)
+      const multihashDigest = digestDecode(multihash)
+      const checksum = base64pad.baseEncode(multihashDigest.digest)
+      const cmd = new PutObjectCommand({
+        Key: `${encodedMultihash}/${encodedMultihash}.blob`,
+        Bucket: bucketName,
+        ChecksumSHA256: checksum,
+        ContentLength: size,
+      })
+      const expiresIn = 60 * 60 * 24 // 1 day
```

Comment:
Nit: Should we expose this as an optional argument ? Seems like something we may want to be able to change in some cases.

Reply:
yes, good call! I will expose this in interface and here